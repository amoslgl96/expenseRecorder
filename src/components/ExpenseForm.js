

import React from 'react';

//import the moment library 

import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

//remember we have webpack to convert this .css file
//into <style> tag into our .html page 
//you can find this .css file in the node_modules

//for the singleDatePicker component, we can find
//the props that we can pass into the component for usage

/*
with the datepicker,
we need to learn how to use the react-dates library
to generate moment instances from the moment library



We can also configure datepicker
to show how many number of months to be on display.

check the documentaton for the necessary props to show:

like numberOfMonths -> by default, 2 months is shown at once 
when focus is true.
we can set it to be 1 only 


we also want to be able to select months before today.


*/

export default class ExpenseForm extends React.Component {

    constructor(props)
    {
        //initialize the props
        super(props);
        //using ternary operator, we check if this.props.expense exist, if it does
        //populate the current values into the state

        //amount and createdAt, need to convert them to string values
        //b4 populating the state
        this.state={
            description: props.expense?props.expense.description:'',
            note: props.expense?props.expense.note:'',
            amount:props.expense?(props.expense.amount/100).toString():'',
            createdAt: props.expense? moment(props.expense.createdAt):moment(),
            calenderFocused: false,
            error:''
        }
    }


    onDescriptionChange=(e)=>{
        const description=e.target.value;
        this.setState(()=>({
            description
        }))
    }

    onNoteChange=(e)=>{
        const note=e.target.value;
        this.setState(()=>({
            note
        }))
    }




    onAmountChange=(e)=>{
        const amount=e.target.value;
        //the !amount allows us to clear the input, else
        //the regex would prevent our state to be set
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
        {
            this.setState(()=>({
                amount:amount
            }))
        }

    }


    //when you look at the API for SingleDatePicker
    //you would know that the onChange event handlerr receives
    //the moment date as arg much like what our 'e' is for onAmountChange
    onDateChange=(createdAt)=>{
        this.setState(()=>({
            createdAt
        }))
    }


    onFocusChange=({focused})=>{
        //console.log(focused);
        //focused means when the calender opens, and when it closes
        //when it opens = true, when closed= null
        this.setState(()=>({calenderFocused: focused}))
    }

    onSubmit=(e)=>{
        e.preventDefault();

        if(!this.state.description || !this.state.amount)
        {
            //set error state equal to 'Please provide description and amount.'
            this.setState(()=>({
                error: 'Please provide description and amount.'
            }))
        }

        else
        {
            //clear the error 
            this.setState(()=>({
                error: ''
            }))
            //with this onSubmit method passed down by either
            //addexpensePage or EditExpensePage
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount,10) * 100,
                //parseFloat helps to keep the decimal, and converts it from str to float value
                //10 as second arg meaning base 10,  *100 to keep it in terms of cents
                createdAt: this.state.createdAt.valueOf(),
                //for createdAt -> we also need to parse it, because we need a regular number
                //that represents its value instead of moment instance
                //using the moment obj method valueOf -> we can convert to number (Epoch number)
                note:this.state.note

            })
        }
    }

    render()
    {
        return (
                <form className="form" onSubmit={this.onSubmit}>
                    {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <input
                        type='text'
                        className='text-input'
                        placeholder='Description'
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    {/* while input type='number' already allows us to prevent
                    userinput from enter alphabets,
                    we need to also find a way to prevent users from entering
                    numbers with more than 2 decimal points. 
                    Hence, to get this done, we should just change the type to 'text'
                    so we can do our own validation code
                    */}
                    <input
                        type='text'
                        className='text-input'
                        placeholder='Amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}

                    />
                    {/*for the createAt which is
                    a date input, we would use a third-party
                    component for it */}

                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calenderFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={()=>false}
                    />
                    {/* The date props - moment object that represents where we want to start

                        onDateChange-> is a function that we can create that
                        would be called with Moment instance when someone picks a new day



                    */ }

                    <textarea 
                        className='textarea'
                        placeholder='Add a note for your expense'
                        value={this.state.note}
                        onChange={this.onNoteChange}>

                    </textarea>

                    <div>
                        <button className="button">Save Expense</button>
                    </div>
                </form>
        )
    }
}