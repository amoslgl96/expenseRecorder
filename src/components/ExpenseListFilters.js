import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter,sortByDate,sortByAmount,setStartDate,setEndDate} from '../actions/filters'

/*
task: 
set dropdown for sortBy. and update
store according to what is selected
*/

//the dateRangePicker require local state to track focusedInput props
//hence ExpenseListFilters need to change into class based component
import { DateRangePicker } from 'react-dates';


export class ExpenseListFilters extends React.Component {
    
    state={
        calenderFocused: null
    };


    //this function will be called from react-dates library
    //this function will be called with an obj
    //this obj will have a startDate and endDate
    onDatesChange=({startDate, endDate})=>{
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    //this method would receive the new value as arg automatically
    onFocusChange=(calenderFocused)=>{
        this.setState(()=>({calenderFocused}))
    }

    onTextChange=(e)=>{
        this.props.setTextFilter(e.target.value);
    }


    onSortChange=(e)=>{
        let selectedOption=e.target.value;
        switch(selectedOption)
        {
            case 'date':
                this.props.sortByDate();
                break;
            case 'amount':
                this.props.sortByAmount();
                break;

        }
    }


    render() {
        return (
            <div>
                <input 
                value={this.props.filters.text} 
                onChange={this.onTextChange}
                type='text'/>
        
                <select 
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange
                    }>
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>

                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calenderFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                    showClearDates={true}
                />
            </div>
        )
    }
}

//regular component 





//function to pass redux store data to ExpenseListFilters
const mapStateToProps=(state)=>({
    filters:state.filters
    //implicit dispatch() is also passed down as props
})



const mapDispatchToProps=(dispatch)=>({
    setStartDate:(startDate)=>{
        dispatch(setStartDate(startDate));
    },

    setEndDate:(endDate)=>{
        dispatch(setEndDate(endDate));
    },

    setTextFilter:(text)=>{
        dispatch(setTextFilter(text))
    },

    sortByDate:()=>{
        dispatch(sortByDate());
    },

    sortByAmount:()=>{
        dispatch(sortByAmount());
    }


})




//export default ExpenseListFilters;
export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters);
