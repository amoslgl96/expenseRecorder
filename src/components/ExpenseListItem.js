
import React from 'react';
import { Link } from 'react-router-dom'

import moment from 'moment';

import numeral from 'numeral';
//wiring up edit_expense
//first set up a Link to edit-expense page with the expense' id


const ExpenseListItem=({description,amount,createdAt,id})=>(
    <div>
        <Link to={`/edit/${id}`}><h3>{description}</h3></Link>
        <p>
        {numeral(amount/100).format('$0,0.00')} 
        - 
        {moment(createdAt).format("MMMM Do, YYYY")}
        </p>
       
    </div>
)
//export const removeExpense = ({ id } = {}) => ({


export default ExpenseListItem;
//if connect() does not specify any arg
// it just means there are no state values passed into
// expenselistitem, but dispatch is still passed in