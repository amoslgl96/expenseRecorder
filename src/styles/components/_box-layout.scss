

//url takes reference to the directory AT the browser
//meaning root dir is at public folder 

.box-layout{
    background: url('/images/bg.jpg');
    //by default the background image
    //is not going to fill up max width and height
    //so we need to manually set that 

    background-size:cover;
    //cover means the image would stretch along to cover entire container 

    height: 100vh;
    //vh -> viewport height , 100 = 100 %
    //The viewport is the user's visible area of a web page. 
    width: 100vw;


    //to center the elements, we use flexbox 
    display:flex;
    justify-content: center; //center horizontally
    align-items:center; //center vertically

}


.box-layout__box{
    background:fade-out(white,0.2);
    border-radius: 3px;
    padding: $l-size $m-size;
    text-align: center;
    width: 25rem;

}


//line-height spacing between lines 

.box-layout__title {
    margin: 0 0 $m-size 0;
    line-height: 1;
}
