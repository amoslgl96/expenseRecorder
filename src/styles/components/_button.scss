
.button {
    background: $blue;
    border: none;
    color: white;
    display:inline-block;
    font-size: $font-size-large;
    font-weight: 300;
    line-height: 1; //make button box skinnier
    padding: $s-size;
    text-decoration:none;
}
/*
display: inline-block allows us to set a width and height on the element.

Also, with display: inline-block, the top and bottom margins/paddings are respected

by default our links are display: inline
this means we would not be able to have our padding and margin
*/



// link modifier, for logout button , 
// using -- allows us to indiicate modify the button
// on top of the styles available from .button
// className would be "button button--link"
.button--link {
    background:none;
}

.button--secondary {
    background: #888888;
}