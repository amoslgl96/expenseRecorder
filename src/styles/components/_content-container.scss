//this is where we define the max-width container 
//that centers everything nicely

.content-container{
    //to ensure that it is align with proper space on left & right
    margin: 0 auto;
    padding:0 $m-size;
    max-width: 80rem;
}