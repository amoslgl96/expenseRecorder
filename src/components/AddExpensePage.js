import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses'



//change to class-based component, to abstract
// the addExpense out from the component 
// using mapDispatchToProps , another functionality 
// other than mapStateToProps.

export class AddExpensePage extends React.Component{
  onSubmit=(expense)=>{
    this.props.addExpense(expense);
    this.props.history.push('/');
  }
  render()
  {
    return(
       <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}/>
      </div>
        )
  }
}

// const AddExpensePage = (props) => (
//   <div>
//     <h1>Add Expense</h1>
//     <ExpenseForm
//       onSubmit={(expense)=>{
//         props.dispatch(addExpense(expense));
//         props.history.push('/');//this would not go thru the server/page refresh but via the BrowserRouter
//         }}/>
//   </div>
// );

const mapDispatchToProps=(dispatch)=>{
  return {
    addExpense:(expense)=>dispatch(addExpense(expense))
  }
}


export default connect(undefined,mapDispatchToProps)(AddExpensePage);
