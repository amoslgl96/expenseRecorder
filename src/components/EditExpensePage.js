import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, startRemoveExpense} from '../actions/expenses';

//with the expense id, we can retrieve the global store
//by connect this component page to the store
//and find the particular expense 



export class EditExpensePage extends React.Component{

  onSubmit=(expense)=>{
    this.props.editExpense(expense);
    this.props.history.push('/')
  }


  onRemove=()=>{
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push('/')

  }
  
  render(){
  return (
    <div>
      <ExpenseForm
        expense={this.props.expense}
        onSubmit={this.onSubmit}/>

        
      <button onClick={this.onRemove}>Remove Expense</button>
    </div>

  );
  }

};


//array.find() -> if returns true, it would stop as it has
//found the particular item it is looking for

const mapStateToProps=(state,props)=>({
  expense: state.expenses.find((expense)=>{
    return expense.id===props.match.params.id
  })
})


const mapDispatchToProps=(dispatch)=>({
  editExpense: (expense)=>dispatch(editExpense(expense.id,expense)),
  startRemoveExpense: (id)=>dispatch(startRemoveExpense(id))
})

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);


/*
Our react-router renders the HOC by calling connect()(editExpensePage)


passes props into the HOC,

hence for mapStateToProps,
the second arg props, is what the current props are,
or what props are being passed into editexpensepage,
this props is passed via the router.

using this current props,
we can work with the global state to retrieve
the correct expense to pass over to EditExpensePage
component as a prop.

*/
