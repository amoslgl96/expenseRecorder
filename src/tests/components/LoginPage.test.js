
import { shallow } from 'enzyme';

import React from 'react';

import {LoginPage} from '../../components/LoginPage';


test("Should render login page properly",()=>{

    const wrapper=shallow(<LoginPage startLogin={()=>{}}/>);
    expect(wrapper).toMatchSnapshot();
})

//test if startLogin gets called when login is pressed
test("Should call startLogin on button click ",
()=>{
    const startLogin = jest.fn();
    const wrapper= shallow(<LoginPage startLogin={startLogin}/>);
    wrapper.find('button').simulate('click');
    expect(startLogin).toHaveBeenCalled();
})
