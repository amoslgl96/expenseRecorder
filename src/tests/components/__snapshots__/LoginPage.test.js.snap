// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render login page properly 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
